src/topology/basic.lean

```diff
@@ -596,14 +596,23 @@ begin
 end
 
 section lim
-variables [nonempty α]
 
-/-- If `f` is a filter, then `lim f` is a limit of the filter, if it exists. -/
-noncomputable def lim (f : filter α) : α := epsilon $ λa, f ≤ 𝓝 a
+/-- If `f` is a filter, then `Lim f` is a limit of the filter, if it exists. -/
+noncomputable def Lim [nonempty α] (f : filter α) : α := epsilon $ λa, f ≤ 𝓝 a
 
-lemma lim_spec {f : filter α} (h : ∃a, f ≤ 𝓝 a) : f ≤ 𝓝 (lim f) := epsilon_spec h
-end lim
+/-- If `f` is a filter in `β` and `g : β → α` is a function, then `lim f` is a limit of `g` at `f`,
+if it exists. -/
+noncomputable def lim [nonempty α] (f : filter β) (g : β → α) : α :=
+Lim (f.map g)
+
+lemma Lim_spec {ha : nonempty α} {f : filter α} (h : ∃a, f ≤ 𝓝 a) : f ≤ 𝓝 (Lim f) :=
```

Comment:
Why did you change from `[nonempty α]` to `{nonempty α}` here and below? My guess is that you consider that `Lim f` already contains a nonemptiness witness, so this assumption will be filled automatically. But suppose that you want to use `continuous.lim_eq` below to rewrite from right to left. Then won't that be a problem? On the other hand, you have kept the typeclass argument in `cauchy.le_nhds_Lim`, where I have the impression that the implicit argument couldn't create a problem like that in rewriting (or maybe it could if one applies `le_trans (cauchy.le_nhds_Lim hf) _`?) Do we have a good rule here?

Reply:
Indeed, with `{h : nonempty α}` we can avoid `@`, e.g., in `dense_embedding` and `uniform_embedding`.

@gebner I had an impression that now Lean can fill in these instances by unification. Does it fail because `nonempty` is a `Prop`? If not, why?

Reply:
I see where the confusion comes from.  Filling in instance-implicit arguments by unification is only supported in type-class search.  The elaborator still tries to synthesize the instance arguments, even if they could be inferred via unification.  (That is, it only works for `instance`, not for `lemma`.)

Reply:
I changed lemmas like `lim_eq` to use a `[nonempty α]` instance derived from other arguments. This way Lean can unify the statement with any other instance and doesn't need a `nonempty` instance to apply a lemma.

Reply:
You could also do this in `Lim_spec` and `lim_spec`, removing the nonemptiness assumption and extracting the witness in the statement from `h`.